plugins/pulsar-records-storage/src/main/java/com/github/bsideup/liiklus/pulsar/PulsarRecordsStorage.java

```diff
@@ -0,0 +1,196 @@
+
+@RequiredArgsConstructor
+@FieldDefaults(makeFinal = true, level = AccessLevel.PRIVATE)
+public class PulsarRecordsStorage implements RecordsStorage {
+
+    public static MessageId fromOffset(long offset) {
+        return new MessageIdImpl(offset >>> 28, offset & 0x0F_FF_FF_FFL, -1);
+    }
+
+    public static long toOffset(MessageId messageId) {
+        MessageIdImpl msgId = (MessageIdImpl) messageId;
+        // Combine ledger id and entry id to form offset
+        // Use less than 32 bits to represent entry id since it will get
+        // rolled over way before overflowing the max int range
+        return (msgId.getLedgerId() << 28) | msgId.getEntryId();
+    }
+
+    PulsarClient pulsarClient;
+
+    ConcurrentMap<String, Mono<Producer<byte[]>>> producers = new ConcurrentHashMap<>();
+
+    @Override
+    public CompletionStage<OffsetInfo> publish(Envelope envelope) {
+        val topic = envelope.getTopic();
+        return producers
+                .computeIfAbsent(topic, __ -> {
+                    return Mono.fromCompletionStage(
+                            pulsarClient.newProducer()
+                                    .topic(topic)
+                                    .hashingScheme(HashingScheme.Murmur3_32Hash)
+                                    .enableBatching(false)
+                                    .createAsync()
+                    ).cache();
+                })
+                .flatMap(producer -> {
+                    val valueBytes = new byte[envelope.getValue().remaining()];
+                    envelope.getValue().duplicate().get(valueBytes);
+                    val future = producer.newMessage()
+                            .key(StandardCharsets.UTF_8.decode(envelope.getKey().duplicate()).toString())
+                            .value(valueBytes)
+                            .sendAsync();
+
+                    return Mono.fromCompletionStage(future)
+                            .cast(MessageIdImpl.class)
+                            .map(it -> new OffsetInfo(
+                                    topic,
+                                    it.getPartitionIndex(),
+                                    toOffset(it)
+                            ));
+                })
+                .toFuture();
+    }
+
+    @Override
+    public Subscription subscribe(String topic, String groupName, Optional<String> autoOffsetReset) {
+        return new PulsarSubscription(topic, groupName, autoOffsetReset);
+    }
+
+    @Value
+    private class PulsarSubscription implements Subscription {
+
+        String topic;
+
+        String groupName;
+
+        Optional<String> autoOffsetReset;
+
+        @Override
+        public Publisher<Stream<? extends PartitionSource>> getPublisher(
+                Supplier<CompletionStage<Map<Integer, Long>>> offsetsProvider
+        ) {
+            return Mono
+                    .defer(() -> Mono.fromCompletionStage(((PulsarClientImpl) pulsarClient).getNumberOfPartitions(topic)))
```

Comment:
why so many casts? :)

Reply:
Ask Pulsar guys :D

Reply:
For some reason this method is not a part of the interface

Reply:
I just saw this what is the project trying to do ?

Reply:
> For some reason this method is not a part of the interface

Pulsar tries to hide the "partitions" complexity from the users in its unified messaging model. Applications don't have to know the "partitions" under the hood. That's why it was not part of the interface.

alternatively, you can use PulsarAdmin for querying the partition metadata : https://github.com/apache/pulsar/blob/master/pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/Topics.java#L263

Reply:
@aahmed-se we're using Liiklus at https://www.vivy.com as our CQRS gateway. Currently we run it with Kafka but plan to deploy Pulsar next to Kafka and move some of our workloads as a first step (with kafka-to-pulsar replication), this is why I added Pulsar support to Liiklus :)

@sijie I see. Thanks for explanation! I was using pulsar-kafka-compat as a ref, this is where I found this code. I didn't want to use PulsarAdmin given its "admin" functionality (and also another transport)